### PR TITLE
First-pass at aes-gcm support for mscng and openssl backends

### PR DESCRIPTION
**Pull request by snargit**
First-pass at aes-gcm support for mscng and openssl backends. No tests yet.

### CODE REVIEW

File: src/mscng/ciphers.c [hunk 1]
```diff
@@ -81,10 +88,23 @@ xmlSecMSCngBlockCipherCheckId(xmlSecTransformPtr transform) {
     return(0);
 }
 
+static int xmlSecMSCngBlockCipherIsCBC(const xmlChar *cipherName)
```
Comment (lsh123) [inline src/mscng/ciphers.c]: I really don't like this idea of looking up a magic string in the cipher name here. Please consider other options:
- Adding a flag in ctx and setting it from xmlSecMSCngBlockCipherInitialize();
- Setting up BCRYPT_CHAIN_MODE_CBC/BCRYPT_CHAIN_MODE_GCM directly in the ctx from xmlSecMSCngBlockCipherInitialize();
- Using transform ID instead

File: src/mscng/ciphers.c [hunk 2]
```diff
@@ -230,22 +293,21 @@ xmlSecMSCngBlockCipherSetKey(xmlSecTransformPtr transform, xmlSecKeyPtr key) {
     bufData = xmlSecBufferGetData(buffer);
     xmlSecAssert2(bufData != NULL, -1);
 
-    /* allocate the key object */
+    dwKeyObjectLength = 0;
```
Comment (lsh123) [inline src/mscng/ciphers.c]: did you remove the comment on purpose?

File: src/mscng/ciphers.c [hunk 3]
```diff
@@ -230,22 +293,21 @@ xmlSecMSCngBlockCipherSetKey(xmlSecTransformPtr transform, xmlSecKeyPtr key) {
...
+    if (status != STATUS_SUCCESS) {
         xmlSecMSCngNtError("BCryptGetProperty",
             xmlSecTransformGetName(transform), status);
         return(-1);
     }
 
-    ctx->pbKeyObject = xmlMalloc(ctx->cbKeyObject);
-    if(ctx->pbKeyObject == NULL) {
-        xmlSecMallocError(ctx->cbKeyObject, NULL);
+    ctx->pbKeyObject = xmlMalloc(dwKeyObjectLength);
```
Comment (lsh123) [inline src/mscng/ciphers.c]: Thank you for fixing this, the old way is hackish. Could you please also add xmlSecAssert(ctx->pbKeyObject == NULL) at the top of the function?

File: src/mscng/ciphers.c [hunk 4]
```diff
@@ -255,50 +317,52 @@ xmlSecMSCngBlockCipherSetKey(xmlSecTransformPtr transform, xmlSecKeyPtr key) {
     if(ret < 0) {
         xmlSecInternalError2("xmlSecBufferInitialize",
             xmlSecTransformGetName(transform), "size=%d", blobHeaderLen);
+        xmlFree(ctx->pbKeyObject);
```
Comment (lsh123) [inline src/mscng/ciphers.c]: You should not be doing this here -- it will result in double free in Finalize()

File: src/mscng/ciphers.c [hunk 5]
```diff
@@ -255,50 +317,52 @@ xmlSecMSCngBlockCipherSetKey(xmlSecTransformPtr transform, xmlSecKeyPtr key) {
...
-xmlSecMSCngBlockCipherCtxInit(xmlSecMSCngBlockCipherCtxPtr ctx,
+static int xmlSecMSCngCBCBlockCipherCtxInit(xmlSecMSCngBlockCipherCtxPtr ctx,
         xmlSecBufferPtr in, xmlSecBufferPtr out, int encrypt,
         const xmlChar* cipherName, xmlSecTransformCtxPtr transformCtx) {
+
     DWORD dwBlockLen, dwBlockLenLen;
     NTSTATUS status;
     int ret;
 
-    xmlSecAssert2(ctx != NULL, -1);
```
Comment (lsh123) [inline src/mscng/ciphers.c]: Why did you remove these asserts?

File: src/mscng/ciphers.c [hunk 6]
```diff
@@ -255,50 +317,52 @@ xmlSecMSCngBlockCipherSetKey(xmlSecTransformPtr transform, xmlSecKeyPtr key) {
...
-    xmlSecAssert2(ctx->hKey != 0, -1);
```
Comment (lsh123) [inline src/mscng/ciphers.c]: Why did you remove these asserts?

File: src/mscng/ciphers.c [hunk 7]
```diff
@@ -308,22 +372,25 @@ xmlSecMSCngBlockCipherCtxInit(xmlSecMSCngBlockCipherCtxPtr ctx,
...
-    if(status != STATUS_SUCCESS) {
+    if (status != STATUS_SUCCESS) {
         xmlSecMSCngNtError("BCryptGetProperty", cipherName, status);
         return(-1);
     }
 
     xmlSecAssert2(dwBlockLen > 0, -1);
-    if(encrypt) {
+
+    ctx->cbIV = dwBlockLen;
```
Comment (lsh123) [inline src/mscng/ciphers.c]: Why do you need this?

File: src/mscng/ciphers.c [hunk 8]
```diff
@@ -375,8 +444,198 @@ xmlSecMSCngBlockCipherCtxInit(xmlSecMSCngBlockCipherCtxPtr ctx,
...
+        const xmlChar* cipherName, xmlSecTransformCtxPtr transformCtx) {
+
+    NTSTATUS status;
+    int ret;
+    xmlSecByte *bufferPtr;
+    xmlSecSize bufferSize;
+    DWORD dwBlockLen, bytesRead;
+    BCRYPT_AUTH_TAG_LENGTHS_STRUCT authTagLengths;
+
+    /* unreferenced parameter */
```
Comment (lsh123) [inline src/mscng/ciphers.c]: Please add asserts to validate your assumptions about input parameters

File: src/mscng/ciphers.c [hunk 9]
```diff
@@ -375,8 +444,198 @@ xmlSecMSCngBlockCipherCtxInit(xmlSecMSCngBlockCipherCtxPtr ctx,
...
+    (void)transformCtx;
+
+    ctx->authInfo.pbNonce = xmlMalloc(12);
+    if (ctx->authInfo.pbNonce == NULL) {
+        xmlSecMallocError(12, cipherName);
```
Comment (lsh123) [inline src/mscng/ciphers.c]: Why size 12? this probably should be a constant

File: src/mscng/ciphers.c [hunk 10]
```diff
@@ -375,8 +444,198 @@ xmlSecMSCngBlockCipherCtxInit(xmlSecMSCngBlockCipherCtxPtr ctx,
...
+        return(-1);
+    }
+    ctx->authInfo.cbNonce = 12;
+
+    /* Tag length is 128 bits */
+    /* See http://www.w3.org/TR/xmlenc-core1/#sec-AES-GCM */
+    ctx->authInfo.pbTag = xmlMalloc(16);
```
Comment (lsh123) [inline src/mscng/ciphers.c]: Same, 16 should be a define

File: src/mscng/ciphers.c [hunk 11]
```diff
@@ -375,8 +444,198 @@ xmlSecMSCngBlockCipherCtxInit(xmlSecMSCngBlockCipherCtxPtr ctx,
...
+
+        /* iv len == block len */
+        status = BCryptGetProperty(ctx->hAlg,
+                                   BCRYPT_BLOCK_LENGTH,
+                                   (PUCHAR)&dwBlockLen,
+                                   sizeof(dwBlockLen),
+                                   &bytesRead,
+                                   0);
+        if (status != STATUS_SUCCESS) {
+            xmlFree(ctx->authInfo.pbTag);
```
Comment (lsh123) [inline src/mscng/ciphers.c]: You don't need to do it, it will be freed with ctx in finalize()

File: src/mscng/ciphers.c [hunk 12]
```diff
@@ -375,8 +444,198 @@ xmlSecMSCngBlockCipherCtxInit(xmlSecMSCngBlockCipherCtxPtr ctx,
...
+            xmlFree(ctx->authInfo.pbNonce);
+            xmlSecMSCngNtError("BCryptGetProperty", cipherName, status);
+            return(-1);
+        }
+        if (ctx->pbIV == NULL) {
+            ctx->pbIV = xmlMalloc(dwBlockLen);
+        }
+        if (ctx->pbIV == NULL) {
+            xmlFree(ctx->authInfo.pbTag);
```
Comment (lsh123) [inline src/mscng/ciphers.c]: Same

File: src/mscng/ciphers.c [hunk 13]
```diff
@@ -375,8 +444,198 @@ xmlSecMSCngBlockCipherCtxInit(xmlSecMSCngBlockCipherCtxPtr ctx,
...
+
+        /* Setup an empty MAC context if we're chaining calls */
+        status = BCryptGetProperty(ctx->hAlg,
+            BCRYPT_AUTH_TAG_LENGTH,
+            (PUCHAR)&authTagLengths,
+            (ULONG)sizeof(authTagLengths),
+            &bytesRead,
+            0);
+        if (status != STATUS_SUCCESS) {
+            xmlFree(ctx->pbIV);
```
Comment (lsh123) [inline src/mscng/ciphers.c]: Same

File: src/mscng/ciphers.c [hunk 14]
```diff
@@ -375,8 +444,198 @@ xmlSecMSCngBlockCipherCtxInit(xmlSecMSCngBlockCipherCtxPtr ctx,
...
+            ctx->pbIV = NULL;
+            xmlFree(ctx->authInfo.pbTag);
+            xmlFree(ctx->authInfo.pbNonce);
+            xmlSecMSCngNtError("BCryptGetProperty", cipherName, status);
+            return(-1);
+        }
+
+        ctx->authInfo.pbMacContext = xmlMalloc(authTagLengths.dwMaxLength);
+        if (ctx->authInfo.pbMacContext == NULL) {
+            xmlFree(ctx->pbIV);
```
Comment (lsh123) [inline src/mscng/ciphers.c]: same

File: src/mscng/ciphers.c [hunk 15]
```diff
@@ -375,8 +444,198 @@ xmlSecMSCngBlockCipherCtxInit(xmlSecMSCngBlockCipherCtxPtr ctx,
...
+
+        bufferPtr = xmlSecBufferGetData(in);
+
+        xmlSecAssert2(bufferPtr != NULL, -1);
+
+        /* set nonce */
+        memcpy(ctx->authInfo.pbNonce, bufferPtr, 12);
+
+        /* remove nonce from input */
+        ret = xmlSecBufferRemoveHead(in, 12);
```
Comment (lsh123) [inline src/mscng/ciphers.c]: 12 needs to be a constant

File: src/mscng/ciphers.c [hunk 16]
```diff
@@ -513,11 +925,8 @@ xmlSecMSCngBlockCipherCtxFinal(xmlSecMSCngBlockCipherCtxPtr ctx,
     NTSTATUS status;
     int ret;
 
-    xmlSecAssert2(ctx != NULL, -1);
```
Comment (lsh123) [inline src/mscng/ciphers.c]: Please restore asserts

File: src/mscng/ciphers.c [hunk 17]
```diff
@@ -660,10 +1069,163 @@ xmlSecMSCngBlockCipherCtxFinal(xmlSecMSCngBlockCipherCtxPtr ctx,
...
+xmlSecMSCngGCMBlockCipherCtxFinal(xmlSecMSCngBlockCipherCtxPtr ctx,
+        xmlSecBufferPtr in, xmlSecBufferPtr out, int encrypt,
+        const xmlChar* cipherName, xmlSecTransformCtxPtr transformCtx)
+{
+    xmlSecByte *inBuf, *outBuf;
+    xmlSecSize inBufSize, outBufSize, outLen;
+    DWORD dwCLen;
+    int ret;
+    NTSTATUS status;
+
```
Comment (lsh123) [inline src/mscng/ciphers.c]: Please add asserts to validate assumptions about input params

File: src/mscng/ciphers.c [hunk 18]
```diff
@@ -660,10 +1069,163 @@ xmlSecMSCngBlockCipherCtxFinal(xmlSecMSCngBlockCipherCtxPtr ctx,
...
+    outBufSize = xmlSecBufferGetSize(out);
+    inBufSize = xmlSecBufferGetSize(in);
+    inBuf = xmlSecBufferGetData(in);
+
+    if (encrypt) {
+        ret = xmlSecBufferSetMaxSize(out, outBufSize + inBufSize + 16); /* add space for the tag */
+        if (ret < 0) {
+            xmlSecInternalError2("xmlSecBufferSetMaxSize", cipherName,
+                "size=%d", outBufSize + inBufSize + 16);
```
Comment (lsh123) [inline src/mscng/ciphers.c]: 16 needs to be a define

File: src/mscng/ciphers.c [hunk 19]
```diff
@@ -677,15 +1239,21 @@ xmlSecMSCngBlockCipherExecute(xmlSecTransformPtr transform, int last, xmlSecTran
     xmlSecAssert2(ctx != NULL, -1);
 
     if(transform->status == xmlSecTransformStatusNone) {
+        ctx->pbIV = NULL;
```
Comment (lsh123) [inline src/mscng/ciphers.c]: Should we free the memory?

File: src/openssl/ciphers.c [hunk 20]
```diff
@@ -71,6 +72,20 @@ static int      xmlSecOpenSSLEvpBlockCipherCtxFinal     (xmlSecOpenSSLEvpBlockCi
...
                                                          const xmlChar* cipherName,
                                                          xmlSecTransformCtxPtr transformCtx);
+#ifndef XMLSEC_NO_AES
+static int      xmlSecOpenSSLIsCBCMode                  (const xmlChar *cipherName)
+{
+    const xmlChar* chainingMode;
+
+    /* cbc or gcm mode */
+    chainingMode = cipherName + xmlStrlen(cipherName) - 3;
+    if (xmlStrcmp(chainingMode, BAD_CAST"gcm") == 0) {
```
Comment (lsh123) [inline src/openssl/ciphers.c]: Same as above

File: src/openssl/ciphers.c [hunk 21]
```diff
@@ -89,7 +107,17 @@ xmlSecOpenSSLEvpBlockCipherCtxInit(xmlSecOpenSSLEvpBlockCipherCtxPtr ctx,
     xmlSecAssert2(out != NULL, -1);
     xmlSecAssert2(transformCtx != NULL, -1);
 
+#ifndef XMLSEC_NO_AES
+    cbcMode = xmlSecOpenSSLIsCBCMode(cipherName);
+    if(cbcMode) {
+        ivLen = EVP_CIPHER_iv_length(ctx->cipher);
+    } else {
+        ivLen = 12; /* This is the nonce length for GCM mode rather than an IV */
```
Comment (lsh123) [inline src/openssl/ciphers.c]: This should be a define

File: src/openssl/ciphers.c [hunk 22]
```diff
@@ -89,7 +107,17 @@ xmlSecOpenSSLEvpBlockCipherCtxInit(xmlSecOpenSSLEvpBlockCipherCtxPtr ctx,
     xmlSecAssert2(out != NULL, -1);
     xmlSecAssert2(transformCtx != NULL, -1);
 
+#ifndef XMLSEC_NO_AES
```
Comment (lsh123) [inline src/openssl/ciphers.c]: Do we really need infdef here?

File: src/openssl/ciphers.c [hunk 23]
```diff
@@ -144,30 +172,49 @@ xmlSecOpenSSLEvpBlockCipherCtxInit(xmlSecOpenSSLEvpBlockCipherCtxPtr ctx,
      *
      * https://www.w3.org/TR/2002/REC-xmlenc-core-20021210/Overview.html#sec-Alg-Block
      */
+#ifndef XMLSEC_NO_AES
```
Comment (lsh123) [inline src/openssl/ciphers.c]: Do we really need ifndef here?

File: src/openssl/ciphers.c [hunk 24]
```diff
@@ -144,30 +172,49 @@ xmlSecOpenSSLEvpBlockCipherCtxInit(xmlSecOpenSSLEvpBlockCipherCtxPtr ctx,
...
+                                          int inSize,
+                                          xmlSecBufferPtr out,
+                                          const xmlChar* cipherName,
+                                          int final,
+                                          xmlSecByte *tagData) {
     xmlSecByte* outBuf;
     xmlSecSize outSize;
     int blockLen, outLen = 0;
     int ret;
+#ifndef XMLSEC_NO_AES
```
Comment (lsh123) [inline src/openssl/ciphers.c]: Do we really need it here?

File: src/openssl/ciphers.c [hunk 25]
```diff
@@ -177,15 +224,38 @@ xmlSecOpenSSLEvpBlockCipherCtxUpdateBlock(xmlSecOpenSSLEvpBlockCipherCtxPtr ctx,
     xmlSecAssert2(blockLen > 0, -1);
     xmlSecAssert2((inSize % blockLen) == 0, -1);
 
-    /* prepare: ensure we have enough space (+blockLen for final) */
     outSize = xmlSecBufferGetSize(out);
+
+#ifndef XMLSEC_NO_AES
```
Comment (lsh123) [inline src/openssl/ciphers.c]: do we need ifndef?

File: src/openssl/ciphers.c [hunk 26]
```diff
@@ -200,12 +270,37 @@ xmlSecOpenSSLEvpBlockCipherCtxUpdateBlock(xmlSecOpenSSLEvpBlockCipherCtxPtr ctx,
     if(final != 0) {
         int outLen2 = 0;
 
+#ifndef XMLSEC_NO_AES
```
Comment (lsh123) [inline src/openssl/ciphers.c]: same

File: src/openssl/ciphers.c [hunk 27]
```diff
@@ -200,12 +270,37 @@ xmlSecOpenSSLEvpBlockCipherCtxUpdateBlock(xmlSecOpenSSLEvpBlockCipherCtxPtr ctx,
...
+        }
+#endif
+
         ret = EVP_CipherFinal(ctx->cipherCtx, outBuf + outLen, &outLen2);
         if(ret != 1) {
             xmlSecOpenSSLError("EVP_CipherFinal", cipherName);
             return(-1);
         }
 
+#ifndef XMLSEC_NO_AES
```
Comment (lsh123) [inline src/openssl/ciphers.c]: same

File: src/openssl/ciphers.c [hunk 28]
```diff
@@ -223,12 +318,15 @@ xmlSecOpenSSLEvpBlockCipherCtxUpdateBlock(xmlSecOpenSSLEvpBlockCipherCtxPtr ctx,
...
-                                  xmlSecBufferPtr in, xmlSecBufferPtr out,
-                                  const xmlChar* cipherName,
-                                  xmlSecTransformCtxPtr transformCtx) {
+                                     xmlSecBufferPtr in, xmlSecBufferPtr out,
+                                     const xmlChar* cipherName,
+                                     xmlSecTransformCtxPtr transformCtx) {
     xmlSecSize inSize, blockLen, inBlocksLen;
     xmlSecByte* inBuf;
     int ret;
+#ifndef XMLSEC_NO_AES
```
Comment (lsh123) [inline src/openssl/ciphers.c]: same

File: src/openssl/ciphers.c [hunk 29]
```diff
@@ -242,26 +340,65 @@ xmlSecOpenSSLEvpBlockCipherCtxUpdate(xmlSecOpenSSLEvpBlockCipherCtxPtr ctx,
     xmlSecAssert2(blockLen > 0, -1);
 
     inSize = xmlSecBufferGetSize(in);
-    if(inSize <= blockLen) {
+
+#ifndef XMLSEC_NO_AES
```
Comment (lsh123) [inline src/openssl/ciphers.c]: same

File: win32/Makefile.msvc [hunk 30]
```diff
@@ -46,9 +46,9 @@ AUTOCONF = .\configure.txt
 #BINPREFIX = $(PREFIX)\bin
 #INCPREFIX = $(PREFIX)\include
 #LIBPREFIX = $(PREFIX)\lib
-#SOPREFIX = $(PREFIX)\lib
+#SOPREFIX = $(PREFIX)\bin
```
Comment (lsh123) [inline win32/Makefile.msvc]: what is the reason for this change?

File: win32/Makefile.msvc [hunk 31]
```diff
@@ -408,19 +413,19 @@ XMLSEC_MSCNG_CFLAGS     = /D "XMLSEC_CRYPTO_MSCNG" /D "XMLSEC_DEFAULT_CRYPTO=\"m
 # The linker and its options.
 #
 LD 				= link.exe
-LDFLAGS 		= /nologo 
+LDFLAGS 		= /nologo /debug
```
Comment (lsh123) [inline win32/Makefile.msvc]: what is the reason for this change?

File: win32/configure.js [hunk 32]
```diff
@@ -63,7 +63,7 @@ var buildPrefix = ".";
 var buildBinPrefix = "$(PREFIX)\\bin";
 var buildIncPrefix = "$(PREFIX)\\include";
 var buildLibPrefix = "$(PREFIX)\\lib";
-var buildSoPrefix = "$(PREFIX)\\lib";
+var buildSoPrefix = "$(PREFIX)\\bin";
```
Comment (lsh123) [inline win32/configure.js]: same question

File: src/mscng/ciphers.c (hunk 1 above)
Comment (snargit) [reply src/mscng/ciphers.c]: Good point. I completely forgot there was an ID to use.

File: src/mscng/ciphers.c (hunk 2 above)
Comment (snargit) [reply src/mscng/ciphers.c]: No. Will reinstate it

File: src/mscng/ciphers.c (hunk 3 above)
Comment (snargit) [reply src/mscng/ciphers.c]: Will do

File: src/mscng/ciphers.c (hunk 4 above)
Comment (snargit) [reply src/mscng/ciphers.c]: Ah, OK - it wasn't obvious to me that Finalze() was called on failure

**Comment by vmiklos** [comment]
I'm just an observer here, but... could you please add tests covering your new code? E.g. as far as I see NSS could support this in the future, but that's much easier to do if there are existing tests covering the feature itself. Also this way you can make sure that at least your openssl and mscng impl interops with each other.

**Comment by lsh123** [comment]
+1 for adding tests. I am fine with doing it in a separate PR though if it makes it easier for you.

**Comment by snargit** [comment]
I'm planning to add some tests once the PR is in better shape. I think it probably makes sense to do it as a second step

File: win32/Makefile.msvc (hunk 30 above)
Comment (snargit) [reply win32/Makefile.msvc]: Consistency with libxml2. The dlls go in bin on install along with the executables - otherwise, unless you mess with your path, dynamically linked xmlsec.exe won't run from the install location.

File: src/mscng/ciphers.c (hunk 5 above)
Comment (snargit) [reply src/mscng/ciphers.c]: The asserts are now in the common xmlSecMsCngBlockCiperCtxInit which delegates to separate init functions for CBC or GCM mode

File: src/mscng/ciphers.c (hunk 8 above)
Comment (snargit) [reply src/mscng/ciphers.c]: As above, the asserts are all in the common ctx init code that calls this function

File: src/mscng/ciphers.c (hunk 9 above)
Comment (snargit) [reply src/mscng/ciphers.c]: Done

File: src/mscng/ciphers.c (hunk 10 above)
Comment (snargit) [reply src/mscng/ciphers.c]: Done

File: src/mscng/ciphers.c (hunk 11 above)
Comment (snargit) [reply src/mscng/ciphers.c]: Done

File: src/mscng/ciphers.c (hunk 12 above)
Comment (snargit) [reply src/mscng/ciphers.c]: Done

File: src/mscng/ciphers.c (hunk 13 above)
Comment (snargit) [reply src/mscng/ciphers.c]: Done

File: src/mscng/ciphers.c (hunk 14 above)
Comment (snargit) [reply src/mscng/ciphers.c]: Done

File: src/mscng/ciphers.c (hunk 15 above)
Comment (snargit) [reply src/mscng/ciphers.c]: Done

File: src/mscng/ciphers.c (hunk 16 above)
Comment (snargit) [reply src/mscng/ciphers.c]: Again, there is a common Final() that delegates to this (or the GCM mode equivalent) that does the asserts before calling the appropriate one.

File: src/mscng/ciphers.c (hunk 17 above)
Comment (snargit) [reply src/mscng/ciphers.c]: As above, asserts are in the common code.

File: src/mscng/ciphers.c (hunk 18 above)
Comment (snargit) [reply src/mscng/ciphers.c]: Done

File: src/mscng/ciphers.c (hunk 19 above)
Comment (snargit) [reply src/mscng/ciphers.c]: I don't think this is actually needed as the ctx is zeroed out when created. I was just worrying about the case when it can be re-initialised during decryption if there was insufficient data in the buffer on first pass. I think it should be safe to remove this completely.

File: src/mscng/ciphers.c [hunk 33]
```diff
@@ -134,14 +161,32 @@ xmlSecMSCngBlockCipherInitialize(xmlSecTransformPtr transform) {
...
+
     ctx->ctxInitialized = 0;
 
     return(0);
 }
 
 static void
 xmlSecMSCngBlockCipherFinalize(xmlSecTransformPtr transform) {
-    xmlSecMSCngBlockCipherCtxPtr ctx;
+        xmlSecMSCngBlockCipherCtxPtr ctx;
```
[the remark this answers is not in the document]
Comment (snargit) [reply src/mscng/ciphers.c]: Fixed

File: src/mscng/ciphers.c [hunk 34]
```diff
@@ -58,7 +61,6 @@ struct _xmlSecMSCngBlockCipherCtx {
 #define xmlSecMSCngBlockCipherGetCtx(transform) \
     ((xmlSecMSCngBlockCipherCtxPtr)(((unsigned char*)(transform)) + sizeof(xmlSecTransform)))
 
-#ifndef XMLSEC_NO_AES
```
[the remark this answers is not in the document]
Comment (snargit) [reply src/mscng/ciphers.c]: No, this #ifdef seemed to encompass most of the file. There were other nested ones that were identical. I've tried to make them all consistent.

File: win32/configure.js (hunk 32 above)
Comment (snargit) [reply win32/configure.js]: As above

File: src/mscng/ciphers.c (hunk 5 above)
Comment (snargit) [reply src/mscng/ciphers.c]: They're done in common code that calls this. Just trying to ensure there's no duplication.

File: src/mscng/ciphers.c (hunk 6 above)
Comment (snargit) [reply src/mscng/ciphers.c]: As above

File: src/mscng/ciphers.c (hunk 34 above)
[the remark this answers is not in the document]
Comment (lsh123) [reply src/mscng/ciphers.c]: I see you moved it to line 1229. Good catch, thanks!

File: src/mscng/ciphers.c [hunk 35]
```diff
@@ -45,11 +45,15 @@ struct _xmlSecMSCngBlockCipherCtx {
     LPCWSTR pszAlgId;
     BCRYPT_ALG_HANDLE hAlg;
     BCRYPT_KEY_HANDLE hKey;
-    PBYTE pbKeyObject;
-    DWORD cbKeyObject;
+    BCRYPT_AUTHENTICATED_CIPHER_MODE_INFO authInfo;
     PBYTE pbIV;
+    ULONG cbIV;
+    PBYTE pbKeyObject;
+    ULONG cbKeyObject;
```
Comment (lsh123) [inline src/mscng/ciphers.c]: I don't see any usage of cbKeyObject in the new code. Should it be removed?

File: src/mscng/ciphers.c [hunk 36]
```diff
@@ -261,19 +320,21 @@ xmlSecMSCngBlockCipherSetKey(xmlSecTransformPtr transform, xmlSecKeyPtr key) {
     blobHeader = (BCRYPT_KEY_DATA_BLOB_HEADER*)xmlSecBufferGetData(&blob);
     blobHeader->dwMagic = BCRYPT_KEY_DATA_BLOB_MAGIC;
     blobHeader->dwVersion = BCRYPT_KEY_DATA_BLOB_VERSION1;
-    blobHeader->cbKeyData = xmlSecBufferGetSize(buffer);
+    blobHeader->cbKeyData = (ULONG)xmlSecBufferGetSize(buffer);
     memcpy(xmlSecBufferGetData(&blob) + sizeof(BCRYPT_KEY_DATA_BLOB_HEADER),
         bufData, xmlSecBufferGetSize(buffer));
+    xmlSecBufferSetSize(&blob, blobHeaderLen);
```
Comment (lsh123) [inline src/mscng/ciphers.c]: This is redundant. xmlSecBufferInitialize() already sets the blob size to blobHeaderLen.

File: src/mscng/ciphers.c [hunk 37]
```diff
@@ -282,90 +343,78 @@ xmlSecMSCngBlockCipherSetKey(xmlSecTransformPtr transform, xmlSecKeyPtr key) {
         return(-1);
     }
 
+    xmlSecBufferFinalize(&blob);
```
Comment (lsh123) [inline src/mscng/ciphers.c]: Good catch. Thanks.

File: src/mscng/ciphers.c (hunk 16 above)
Comment (lsh123) [reply src/mscng/ciphers.c]: I strongly prefer to have each function check all the assumptions about input parameters. It saves a lot of headache later for a small performance cost.

File: src/mscng/ciphers.c (hunk 6 above)
Comment (lsh123) [reply src/mscng/ciphers.c]: Please restore the input params checks

File: src/mscng/ciphers.c [hunk 38]
```diff
@@ -502,93 +703,219 @@ xmlSecMSCngBlockCipherCtxUpdate(xmlSecMSCngBlockCipherCtxPtr ctx,
...
-    unsigned char* outBuf;
-    DWORD dwBlockLen, dwBlockLenLen, dwCLen;
+
     NTSTATUS status;
+    size_t inSize, outSize;
+    xmlSecByte *inBuf, *outBuf;
+    DWORD dwCLen;
     int ret;
 
+    /* unreferenced parameter */
```
Comment (lsh123) [inline src/mscng/ciphers.c]: Please add assets to validate input parameters

File: src/mscng/ciphers.c [hunk 39]
```diff
@@ -375,74 +424,226 @@ xmlSecMSCngBlockCipherCtxInit(xmlSecMSCngBlockCipherCtxPtr ctx,
...
+        xmlSecBufferPtr in, xmlSecBufferPtr out, int encrypt,
+        const xmlChar* cipherName, xmlSecTransformCtxPtr transformCtx) {
+    size_t inSize, inBlocks, outSize;
+    unsigned char* outBuf;
+    unsigned char* inBuf;
+    DWORD dwCLen;
+    NTSTATUS status;
+    int ret;
+
+    /* unreferenced parameter */
```
Comment (lsh123) [inline src/mscng/ciphers.c]: Please add asserts to validate input parameters

File: src/mscng/ciphers.c [hunk 40]
```diff
@@ -375,74 +424,226 @@ xmlSecMSCngBlockCipherCtxInit(xmlSecMSCngBlockCipherCtxPtr ctx,
...
-    DWORD dwBlockLen, dwBlockLenLen, dwCLen;
+
     NTSTATUS status;
     int ret;
+    xmlSecByte *bufferPtr;
+    xmlSecSize bufferSize;
+    DWORD bytesRead;
+    BCRYPT_AUTH_TAG_LENGTHS_STRUCT authTagLengths;
+
+    /* unreferenced parameter */
```
Comment (lsh123) [inline src/mscng/ciphers.c]: Please add asserts to validate input parameters

File: src/openssl/ciphers.c [hunk 41]
```diff
@@ -277,31 +354,29 @@ xmlSecOpenSSLEvpBlockCipherCtxUpdate(xmlSecOpenSSLEvpBlockCipherCtxPtr ctx,
...
 
-    xmlSecAssert2(ctx != NULL, -1);
-    xmlSecAssert2(ctx->cipher != NULL, -1);
-    xmlSecAssert2(ctx->cipherCtx != NULL, -1);
-    xmlSecAssert2(ctx->keyInitialized != 0, -1);
-    xmlSecAssert2(ctx->ctxInitialized != 0, -1);
-    xmlSecAssert2(in != NULL, -1);
-    xmlSecAssert2(out != NULL, -1);
-    xmlSecAssert2(transformCtx != NULL, -1);
+    /* unreferenced parameter */
```
Comment (lsh123) [inline src/openssl/ciphers.c]: Please add asserts to validate input parameters

File: src/openssl/ciphers.c [hunk 42]
```diff
@@ -398,6 +474,98 @@ xmlSecOpenSSLEvpBlockCipherCtxFinal(xmlSecOpenSSLEvpBlockCipherCtxPtr ctx,
...
+        const xmlChar* cipherName,
+        xmlSecTransformCtxPtr transformCtx)
+{
+    xmlSecSize inSize, outSize;
+    xmlSecByte* inBuf;
+    xmlSecByte* outBuf;
+    xmlSecByte tag[xmlSecOpenSSLAesGcmTagLengthInBytes];
+    int ret;
+
+    /* unreferenced parameter */
```
Comment (lsh123) [inline src/openssl/ciphers.c]: Please add asserts to validate input parameters

File: win32/Makefile.msvc (hunk 30 above)
Comment (lsh123) [reply win32/Makefile.msvc]: Might be easier to isolate the win32 build changes from the rest of the patch.

File: win32/Makefile.msvc (hunk 31 above)
Comment (lsh123) [reply win32/Makefile.msvc]: Might be easier to isolate the win32 makefile changes from the rest of the aes-gcm changes
